integration/ledger/snapshot_test.go

```diff
@@ -0,0 +1,1049 @@
+	var err error
+	testDir, err := workDir("/tmp", "snapshot")
+	Expect(err).NotTo(HaveOccurred())
+
+	client, err := docker.NewClientFromEnv()
+	Expect(err).NotTo(HaveOccurred())
+
+	config := nwo.BasicSolo()
+
+	config.Channels = []*nwo.Channel{
+		{Name: testchannelID, Profile: "TwoOrgsChannel"},
+		{Name: testchannelID2, Profile: "TwoOrgsChannel"},
+		{Name: testchannelID3, Profile: "TwoOrgsChannel"},
+	}
+
+	for _, peer := range config.Peers {
+		peer.Channels = []*nwo.PeerChannel{
+			{Name: testchannelID, Anchor: true},
+			{Name: testchannelID2, Anchor: true},
+			{Name: testchannelID3, Anchor: true},
+		}
+	}
+
+	for _, org := range config.Organizations {
+		if org.Name == "Org1" || org.Name == "Org2" {
+			org.Users = 3
+		}
+	}
+
+	config.Peers = append(
+		config.Peers,
+		&nwo.Peer{
+			Name:         "peer1",
+			Organization: "Org1",
+			Channels:     []*nwo.PeerChannel{},
+		},
+		&nwo.Peer{
+			Name:         "peer1",
+			Organization: "Org2",
+			Channels: []*nwo.PeerChannel{
+				{Name: testchannelID2, Anchor: false},
+			},
+		},
+	)
+
+	n := nwo.New(config, testDir, client, StartPort(), components)
+	n.GenerateConfigTree()
+	n.Bootstrap()
+
+	// only keeps peer0 so we can add other peers back later to test join channel by snapshot
+	peers := []*nwo.Peer{}
+	for _, p := range n.Peers {
+		if p.Name == "peer0" {
+			peers = append(peers, p)
+		}
+	}
+	n.Peers = peers
+
+	setup := &setup{
+		testDir:   testDir,
+		network:   n,
+		peers:     peers,
+		channelID: testchannelID,
+	}
+	Expect(setup.testDir).To(Equal(setup.network.RootDir))
+
+	By("starting broker and orderer")
+	setup.startBrokerAndOrderer()
+
+	By("starting peers")
+	setup.startPeers()
+
+	orderer := n.Orderer("orderer")
+	setup.orderer = orderer
+
+	By("creating and joining testchannel")
+	n.CreateAndJoinChannel(orderer, testchannelID)
+	n.UpdateChannelAnchors(orderer, testchannelID)
+
+	By("verifying membership for testchannel")
+	n.VerifyMembership(n.PeersWithChannel(testchannelID), testchannelID)
+
+	By("creating and joining testchannel2")
+	n.CreateAndJoinChannel(orderer, testchannelID2)
+	n.UpdateChannelAnchors(orderer, testchannelID2)
+
+	By("verifying membership for testchannel2")
+	n.VerifyMembership(n.PeersWithChannel(testchannelID2), testchannelID2)
+
+	By("creating and joining testchannel3")
+	n.CreateAndJoinChannel(orderer, testchannelID3)
+	n.UpdateChannelAnchors(orderer, testchannelID3)
+
+	By("verifying membership for testchannel3")
+	n.VerifyMembership(n.PeersWithChannel(testchannelID3), testchannelID3)
+
+	return setup
+}
+
+func workDir(dir, pattern string) (string, error) {
+	trypath := filepath.Join(dir, pattern)
+	os.RemoveAll(trypath)
+	err := os.Mkdir(trypath, 0700)
+	return trypath, err
+}
+
+// verifySnapshotRequestCmds invokes snapshot commands and verify the expected rersults.
+// At the end, there will be no pending request.
+func verifySnapshotRequestCmds(n *nwo.Network, peer *nwo.Peer, channel string, blockNum int) {
```

Comment:
Do we need so many `By`s in this function... it compromises with the readability. The same comment about other similar functions as well